### PR TITLE
feat(queues): allow to batch pull requests

### PR DESCRIPTION
**Pull request by sileht**
This introduces the batch_size configuration option for queue rules.

In case of failure, we split a batch in X almost equal parts. X is the
value of speculative checks. We remove from the split the last PR of the
batch, because we already tested it

If the first slice of the batch split is mergeable, we merge it. We loop
until one slice doesn't work, The failed slice is split again in smaller
piece until we find the culprit. We, if possible, always create the
maximum (speculative checks value) number of draft pull requests to
checks part of the batch.

To detect failed batch we have to store in redis two new informations:
* the last checks_conclusion of a TrainCar.
* the previously failed TrainCar before it was splitted.

Change-Id: I82c3c6082d435dabdbbfbbb47ea5b43cb020683f

### CODE REVIEW

File: mergify_engine/queue/merge_train.py [hunk 1]
```diff
@@ -591,8 +611,11 @@ async def update_summaries(
             elif conclusion == check_api.Conclusion.PENDING:
                 if will_be_reset:
                     headline = f"✨ Unexpected queue change. The pull request {self._get_user_refs()} will be re-embarked soon. ✨"
-                elif checks_conlusion == check_api.Conclusion.FAILURE:
-                    headline = "🕵️  This combination of pull requests has failed check. Mergify is waiting for other pull requests ahead in the queue to understand which one is responsible for the failure. 🕵️"
+                elif checks_conclusion == check_api.Conclusion.FAILURE:
+                    if not self.has_previous_car_status_succeed():
+                        headline = "🕵️  This combination of pull requests has failed check. Mergify is waiting for other pull requests ahead in the queue to understand which one is responsible for the failure. 🕵️"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:616]:
```suggestion
                        headline = "🕵️  This combination of pull requests has failed checks. Mergify is waiting for other pull requests ahead in the queue to understand which one is responsible for the failure. 🕵️"
```

File: releasenotes/notes/queue-batch-0e6eb62939368608.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+features:
+  - |
+    Add `batch_size` option to queue rule. The new option allows to checks
+    multiple pull requests within each speculative checks. For example, if you have
+    `speculative_checks: 2` and `batch_size: 2`, and we have pull requests #1, #2,
+    #3 queued. Instead of creating draft pull requests for #1 and then #1+#2 to
+    check the meargeability of #1 and #2. We just create one draft pull requests
+    for #1+#2, and if everything is good we will merge #1 and #2 in a
+    row without checking them individualy
```
Comment (jd) [inline releasenotes/notes/queue-batch-0e6eb62939368608.yaml:10]:
```suggestion
    Add ``batch_size`` option to queue rules. The new option allows to check
    multiple pull requests within each speculative check.
```

Don't duplicate the doc here… which is, you should put the rest of the text into the doc, not here.

**Review by jd — DISMISSED** [review]

File: docs/source/actions/queue.rst [hunk 3]
```diff
@@ -277,6 +277,13 @@ A ``queue_rules`` takes the following parameter:
        between 1 and 20.
        See :ref:`speculative checks`.
 
+   * - ``batch_size``
+     - int
+     - 1
+     - |premium plan tag| The maximum number of pull requests per speculative checks in the queue. Must be
+       between 1 and 20.
```
Comment (jd) [inline docs/source/actions/queue.rst:284]:
```suggestion
     - |premium plan tag| The maximum number of pull requests per speculative check in the queue. Must be
       between 1 and 20.
```

File: docs/source/actions/queue.rst [hunk 4]
```diff
@@ -372,14 +379,41 @@ checks <speculative checks>`. This will allow Mergify to trigger multiple runs
 of the CI in parallel.
 
 In the following example, by setting the ``speculative_checks`` option to
-``3``, Mergify will create up to new 2 pull requests to check if the first
+``2``, Mergify will create up to new 2 pull requests to check if the first
```
Comment (jd) [inline docs/source/actions/queue.rst:382]:
```suggestion
``2``, Mergify will create up to 2 new pull requests to check if the first
```

File: docs/source/actions/queue.rst [hunk 5]
```diff
@@ -372,14 +379,41 @@ checks <speculative checks>`. This will allow Mergify to trigger multiple runs
...
+          queue:
+            name: default
+
+Multiple pull requests can be checked within one speculative check by settings
+``batch_size``.
+
+For example, by settings ``speculative_checks: 2`` and ``batch_size: 3``,
+Mergify will create only 1 pull request to check if the first three enqueued
+pull requests are mergeable. And a second one to check the three next enqueued
+pull requests.
```
Comment (jd) [inline docs/source/actions/queue.rst:409]:
```suggestion
For example, by settings ``speculative_checks: 2`` and ``batch_size: 3``,
Mergify will create two pull requests: a first one to check if the first three enqueued pull requests are mergeable, and a second one to check the three next enqueued pull requests.
```

File: mergify_engine/queue/merge_train.py [hunk 6]
```diff
@@ -552,7 +572,7 @@ async def update_summaries(
                     f"The pull request {refs} cannot be merged and has been disembarked"
                 )
             else:
-                tmp_pull_title = f"The pull requests {refs} cannot be merged and have been disembarked"
+                tmp_pull_title = f"The pull requests {refs} cannot be merged and have will be splitted"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:575]:
```suggestion
                tmp_pull_title = f"The pull requests {refs} cannot be merged and will be split"
```

File: mergify_engine/queue/merge_train.py [hunk 7]
```diff
@@ -591,8 +611,11 @@ async def update_summaries(
             elif conclusion == check_api.Conclusion.PENDING:
                 if will_be_reset:
                     headline = f"✨ Unexpected queue change. The pull request {self._get_user_refs()} will be re-embarked soon. ✨"
-                elif checks_conlusion == check_api.Conclusion.FAILURE:
-                    headline = "🕵️  This combination of pull requests has failed check. Mergify is waiting for other pull requests ahead in the queue to understand which one is responsible for the failure. 🕵️"
+                elif checks_conclusion == check_api.Conclusion.FAILURE:
+                    if not self.has_previous_car_status_succeed():
+                        headline = "🕵️  This combination of pull requests has failed checks. Mergify is waiting for other pull requests ahead in the queue to understand which one is responsible for the failure. 🕵️"
+                    else:
+                        headline = "🕵️  This combination of pull requests has failed checks. Mergify will split this batch to understand which one is responsible for the failure. 🕵️"
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:618]:
```suggestion
                    if self.has_previous_car_status_succeed():
                        headline = "🕵️  This combination of pull requests has failed checks. Mergify will split this batch to understand which pull request is responsible for the failure. 🕵️"
                     else:
                        headline = "🕵️  This combination of pull requests has failed checks. Mergify is waiting for other pull requests ahead in the queue to understand which one is responsible for the failure. 🕵️"
```

File: mergify_engine/queue/merge_train.py [hunk 8]
```diff
@@ -671,6 +694,8 @@ async def update_summaries(
         else:
             checks_copy_summary = ""
 
+        # TODO(sileht): Add a section about failure histories
```
Comment (jd) [inline mergify_engine/queue/merge_train.py:697]:
```suggestion
        # TODO(sileht): Add a section about failure history
```

File: mergify_engine/tests/functional/actions/test_queue.py [hunk 9]
```diff
@@ -223,6 +224,149 @@ async def assert_queued():
...
+
+        await self.create_status(tmp_pulls[0])
+        await self.create_status(tmp_pulls[1])
+        await self.run_engine()
+
+        pulls = await self.get_pulls()
+        assert len(pulls) == 0
+
+        await self._assert_cars_contents(q, None, [])
+
```
Comment (jd) [inline mergify_engine/tests/functional/actions/test_queue.py:369]: Could we add a functional scenario for failure with a split?
That'd be really great to cover it here too.

**Review by jd — DISMISSED** [review]

File: mergify_engine/utils.py [hunk 10]
```diff
@@ -220,3 +221,15 @@ async def send_refresh(
     await github_events.filter_and_dispatch(
         redis_cache, redis_stream, "refresh", str(uuid.uuid4()), data
     )
+
+
+T = typing.TypeVar("T")
```
Comment (jd) [inline mergify_engine/utils.py:226]:
```suggestion
_T = typing.TypeVar("_T")
```

**Review by jd — CHANGES_REQUESTED** [review]

**Review by jd — DISMISSED** [review]
Missed my comments I think.

**Review by jd — CHANGES_REQUESTED** [review]
😢

File: mergify_engine/queue/merge_train.py [hunk 11]
```diff
@@ -671,6 +710,8 @@ async def update_summaries(
         else:
             checks_copy_summary = ""
 
+        # TODO(sileht): Add a section about failure history
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/queue/merge_train.py:713]: yes

**Review by jd — APPROVED** [review]

**Pull request merged**